File: src/main/java/com/EmployeesREST/service/EmployeeService.java
package com.EmployeesREST.service;

import com.EmployeesREST.dto.Employee;
import com.EmployeesREST.jms.Producer;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.util.List;

import com.EmployeesREST.exceptions.EmployeeServiceNotFoundException;
import com.EmployeesREST.dao.EmployeeRepository;

@Service
public class EmployeeService {
    private final EmployeeRepository employeeRepository;
    private final Producer producer;

    public EmployeeService(EmployeeRepository employeeRepository, Producer producer) {
        this.employeeRepository = employeeRepository;
        this.producer = producer;
    }

    public Employee getEmployeeById(long employeeID) throws EmployeeServiceNotFoundException {
        return employeeRepository.findById(employeeID).orElseThrow(() ->
                new EmployeeServiceNotFoundException(String.format("Employee with id %d does not exist!", employeeID)));
    }

    public List<Employee> getEmployeesByLastAndFirstName(String lastName, String firstName) throws EmployeeServiceNotFoundException {
        return employeeRepository.findByLastNameContainingIgnoreCaseAndFirstNameContainingIgnoreCase(lastName, firstName);
    }

    public void addEmployee(Employee newEmployee) {
        employeeRepository.save(newEmployee);
    }

    public void deleteEmployee(Long employeeToDeleteId) throws EmployeeServiceNotFoundException {
        employeeRepository.deleteEmployeesByEmployeeID(employeeToDeleteId);
    }

    @Transactional
    public void updateEmployee(Employee updatedEmployee, long employeeId) throws EmployeeServiceNotFoundException {
        employeeRepository.findById(employeeId).orElseThrow(() ->
                new EmployeeServiceNotFoundException(String.format("Employee with id %d does not exist!", employeeId)));

        employeeRepository.save(Employee.builder().employeeID(employeeId).firstName(updatedEmployee.getFirstName())
                .lastName(updatedEmployee.getLastName()).dateOfBirth(updatedEmployee.getDateOfBirth()).gender(updatedEmployee
                        .getGender()).departmentID(updatedEmployee.getDepartmentID()).jobTitle(updatedEmployee.getJobTitle()).build());
    }

    public void addEmployeeUsingJMS(Employee employee){
        producer.addEmployee(employee);
    }
}

File: src/main/korlin/com/employees/service/EmployeeService.kt
package com.employees.service

import com.employees.dao.EmployeeRepository
import com.employees.dto.Employee
import com.employees.exceptions.EmployeeServiceNotFoundException
import com.employees.jms.EmployeeJmsProducer
import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
class EmployeeService(employeeRepository: EmployeeRepository, producer: EmployeeJmsProducer) {
    private val employeeRepository: EmployeeRepository
    private val producer: EmployeeJmsProducer

    init {
        this.employeeRepository = employeeRepository
        this.producer = producer
    }

    @Throws(EmployeeServiceNotFoundException::class)
    fun getEmployeeById(employeeID: Long): Employee? {
        return employeeRepository.findById(employeeID).orElseThrow {
            EmployeeServiceNotFoundException(
                String.format(
                    "Employee with id %d does not exist!",
                    employeeID
                )
            )
        }
    }

    @Throws(EmployeeServiceNotFoundException::class)
    fun getEmployeesByLastAndFirstName(lastName: String?, firstName: String?): List<Employee?>? {
        return employeeRepository.findByLastNameContainingIgnoreCaseAndFirstNameContainingIgnoreCase(
            lastName,
            firstName
        )
    }

    fun addEmployee(newEmployee: Employee?) {
        employeeRepository.save(newEmployee!!)
    }

    @Throws(EmployeeServiceNotFoundException::class)
    fun deleteEmployee(employeeToDeleteId: Long?) {
        employeeRepository.deleteEmployeesByEmployeeID(employeeToDeleteId)
    }

    @Transactional
    @Throws(EmployeeServiceNotFoundException::class)
    fun updateEmployee(updatedEmployee: Employee, employeeId: Long) {
        employeeRepository.findById(employeeId).orElseThrow {
            EmployeeServiceNotFoundException(
                String.format(
                    "Employee with id %d does not exist!",
                    employeeId
                )
            )
        }
        employeeRepository.save(Employee(employeeId, updatedEmployee.firstName, updatedEmployee.lastName, updatedEmployee.gender, updatedEmployee.dateOfBirth,
        updatedEmployee.jobTitle, updatedEmployee.departmentID))
    }

    fun addEmployeeUsingJMS(employee: Employee?) {
        producer.addEmployee(employee)
    }
}